### PR TITLE
refactor: make `List.length_eq_countP_add_countP` avoid Prop

### PR DESCRIPTION
**Pull request by BoltonBailey**
Changes this lemma to use boolean not rather than propositional not.

### CODE REVIEW

**Comment by BoltonBailey** [comment]
Is this repo accepting PRs that could break downstream? Is there anything I need to do to make sure mathlib doesn't break here?

**Comment by BoltonBailey** [comment]
WIP

**Comment by Shreyas4991** [comment]
> Is this repo accepting PRs that could break downstream? Is there anything I need to do to make sure mathlib doesn't break here?

There's a [bot](https://github.com/leanprover-community-mathlib4-bot) that will add a merge-conflict label if this is an issue.

File: Batteries/Data/List/Count.lean [hunk 1]
```diff
@@ -44,18 +44,19 @@ protected theorem countP_go_eq_add (l) : countP.go p l n = n + countP.go p l 0 :
...
       · rw [Nat.add_assoc, Nat.add_comm _ 1, Nat.add_assoc]
-      · simp only [h, not_true_eq_false, decide_False, not_false_eq_true]
+      · simp only [h, not_true_eq_false, decide_False, not_false_eq_true, Bool.not_true,
+          Bool.false_eq_true, not_false_eq_true]
     else
-      rw [countP_cons_of_pos (fun a => ¬p a) _ _, countP_cons_of_neg _ _ h, length, ih]
+      rw [countP_cons_of_pos (fun a => ! p a) _ _, countP_cons_of_neg _ _ h, length, ih]
       · rfl
-      · simp only [h, not_false_eq_true, decide_True]
+      · simp only [h, not_false_eq_true, decide_True, Bool.not_false]
```
Comment (fgdorais) [inline Batteries/Data/List/Count.lean:59]: Simplified proof:
```lean
theorem length_eq_countP_add_countP (l) : length l = countP p l + countP (!p ·) l := by
  induction l with
  | nil => rfl
  | cons x _ _ => cases h : p x <;> simp_arith [*]
```

**Comment by Shreyas4991** [comment]
> Is this repo accepting PRs that could break downstream? Is there anything I need to do to make sure mathlib doesn't break here?

On second thoughts, I might help to make one of those testing PRs in mathlib. I was asked to make this for another PR. Im such a PR, one basically changes the repository's batteries dependency to one's own fork and PR branch in Mathlib's lakefile and lake manifest

**Comment by fgdorais** [comment]
Upstreamed to core.

**Pull request closed without merging**